prow/gangway/gangway.proto

```diff
@@ -0,0 +1,129 @@
+syntax = "proto3";
+
+import "google/api/annotations.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "k8s.io/test-infra/prow/gangway";
+
+service Prow {
+  // FIXME: In the future we can just return a unique token (only), in the same
+  // way that GCB returns immediately with the globally-unique BuildId. That is,
+  // in the future the response will be a union of either the full JobExecution
+  // message or a single JobExecutionToken (string). See
+  // https://docs.google.com/document/d/1v77jp1Nb5C2C2-PdV02SGViO9CyZ9SvNxCPOHyIUQeo/edit#bookmark=id.q68srxklvpt4.
+  rpc CreateJobExecution(CreateJobExecutionRequest) returns (JobExecution) {
+    option (google.api.http) = {
+      post: "/v1/executions/{job_name}"
+      body: "*"  // See
+                 // https://cloud.google.com/endpoints/docs/grpc/transcoding#use_wildcard_in_body
+    };
+  }
+  rpc GetJobExecution(GetJobExecutionRequest) returns (JobExecution) {
+    // Client example:
+    //   curl http://DOMAIN_NAME/v1/executions/1
+    option (google.api.http) = {
+      get: "/v1/executions/{id}"
+    };
+  }
+  rpc ListJobExecutions(ListJobExecutionsRequest) returns (JobExecutions) {
+    // Client example:
+    //   curl
+    //   http://DOMAIN_NAME/v1/executions?job_name=my-prow-job&status=SUCCESS
+    option (google.api.http) = {
+      get: "/v1/executions"
+    };
+  }
+}
+
+message CreateJobExecutionRequest {
+  string job_name = 1;
+  JobExecutionType job_execution_type = 2;
+  Refs refs = 3;
+  PodSpecOptions pod_spec_options = 4;
+}
+
+message PodSpecOptions {
+  map<string, string> envs = 1;
+  map<string, string> labels = 2;
+  map<string, string> annotations = 3;
+}
+
+/* Look up a single Prow Job execution. */
+message GetJobExecutionRequest {
+  string id = 1;
+}
+
+/* Look up all Prow Job executions that match all fields given here. */
+message ListJobExecutionsRequest {
+  string job_name = 1;            // Mapped to URL query parameter `job_name`.
+  JobExecutionStatus status = 2;  // Mapped to URL query parameter `status`.
+}
+
+message JobExecutions {
+  repeated JobExecution job_execution = 1;
+}
+
+message JobExecution {
+  string id = 1;
+  CreateJobExecutionRequest create_job_execution_request = 2;
```

Comment:
Continuing the discussion from [here](https://github.com/kubernetes/test-infra/pull/28036#discussion_r1051328120):
Personally I'm not sure this is the most user friendly and I don't think this really "preserves the object model".
- It doesn't really preserve the object model because the `pod_spec_options` subfield won't be the same as what we triggered with, but rather the concatenation of the configured env/annotations/labels and the additional values from the execution request payload.
- It doesn't seem intuitive for the job name to be included under the execution request object to me. 

Neither of these are technical blockers and I don't feel very strongly about this so we can keep this as is if you'd like, but I just wanted share my thoughts that it might be simpler to just replace this with 3 fields for job_name, execution_type, and refs. Totally up to you.

Reply:
> It doesn't really preserve the object model because the pod_spec_options subfield won't be the same as what we triggered with, but rather the concatenation of the configured env/annotations/labels and the additional values from the execution request payload.

I think this is a good point and I agree with this. Will update.

> It doesn't seem intuitive for the job name to be included under the execution request object to me.

Thinking about this some more, I guess I also agree. The reason why it was included in the request object was simply because it is how clients create request objects (they need to supply a job name).